lib/index.js

```diff
@@ -107,21 +107,40 @@ api.set = async ({
   } else if(type === 'nonce') {
     // FIXME: may need to allow up to N tokens, one per `clientId` if
     // `clientId` is set
-
-    // if there is an existing, unexpired nonce, reuse it, do not generate
-    // a new one to avoid DoS attack of constantly running bcrypt algorithm
+    const maxNonceCount = 5;
+    let existingTokens;
     try {
-      const existingToken = await api.get({actor: null, account, email, type});
-      // keep existing token if it hasn't expired; note that there is no
-      // `challenge` to return because it is not known in the database
-      if(existingToken && new Date() < new Date(existingToken.expires)) {
-        return {type};
-      }
+      // check if nonce exists
+      existingTokens = await api.get({actor: null, account, email, type});
     } catch(e) {
       if(e.name !== 'NotFoundError') {
         throw e;
       }
     }
+    if(existingTokens) {
+      for(const existingToken of existingTokens) {
+        // check if existing token has expired
+        if(existingToken && new Date() > new Date(existingToken.expires)) {
+          try {
+            // the `nonce` has expired, delete it
+            await api.remove({actor, account, type, existingToken});
+          } catch(e) {
+            if(e.name !== 'NotFoundError') {
+              throw e;
+            }
+          }
+        }
+      }
+    }
+    // If existingTokens exceeds maxNonceCount, throw `NotAllowed` error
+    if(existingTokens && existingTokens.length >= maxNonceCount) {
```

Comment:
Looks like the length check needs to account for tokens that were expired above?